File: Data_Structure/segment-tree.py
# verified with:
# https://judge.yosupo.jp/submission/109994
# https://judge.yosupo.jp/submission/109997
# https://atcoder.jp/contests/abc254/submissions/35985460

from typing import TypeVar, Callable, List

T = TypeVar("T")
class segment_tree:
    def __init__(self, _n:int, _fx:Callable[[T,T],T], _ex) -> None:
        self.n = 1
        self.fx = _fx
        self.ex = _ex
        while self.n < _n:
            self.n *= 2
        self.val = [self.ex]*self.n*2
    
    def set_list(self, a:List[T]) -> None:
        for i in range(len(a)):
            self.val[i+self.n] = a[i]
        for i in range(self.n-1,-1,-1):
            self.val[i] = self.fx(self.val[i<<1], self.val[(i<<1)|1])
    
    def add(self, i:int, x:T) -> None:
        i += self.n
        self.val[i] = self.fx(self.val[i], x)
        i >>= 1
        while i:
            self.val[i] = self.fx(self.val[i<<1],self.val[(i<<1)|1])
            i >>= 1
        
    def update(self, i:int, x:T) -> None:
        i += self.n
        self.val[i] = x
        i >>= 1
        while i:
            self.val[i] = self.fx(self.val[i<<1],self.val[(i<<1)|1])
            i >>= 1
    
    def query(self, l:int, r:int) -> T:
        vl = self.ex
        vr = self.ex
        l += self.n
        r += self.n
        while l < r:
            if l&1:
                vl = self.fx(vl,self.val[l])
                l += 1
            if r&1:
                r -= 1
                vr = self.fx(self.val[r],vr)
            l >>= 1
            r >>= 1
        return self.fx(vl,vr)

File: Data_Structure/segment-tree.cpp
// verified with:
// https://judge.yosupo.jp/submission/109990
// https://judge.yosupo.jp/submission/109991
// https://atcoder.jp/contests/abc254/submissions/35985098

#include<bits/stdc++.h>
using namespace std;

// RMQ <int>
// auto fx = [](int x1,int x2) -> int{ return min(x1,x2);}
// int ex = INT_MAX
template <typename T>
struct segment_tree{
  explicit segment_tree(int _n, function<T(T,T)> _fx, T _ex) : fx(_fx), ex(_ex) {
    n = 1;
    while (n < _n) n *= 2;
    val.assign(n*2,ex);
  }

  void set_vec(vector<T> a){
    for(int i = 0; i < (int)a.size(); ++i) val[i+n] = a[i];
    for(int i = n-1; i > 0; --i) val[i] = fx(val[i<<1],val[(i<<1)|1]);
  }

  void add(int i,T x){
    i += n;
    val[i] = fx(val[i],x);
    i >>= 1;
    while(i){
      val[i] = fx(val[i<<1],val[(i<<1)|1]);
      i >>= 1;
    }
  }

  void update(int i,T x){
    i += n;
    val[i] = x;
    i >>= 1;
    while(i){
      val[i] = fx(val[i<<1],val[(i<<1)|1]);
      i >>= 1;
    }
  }

  T query(int l,int r){
    T vl = ex,vr = ex;
    l += n;
    r += n;
    while(l < r) {
      if(l&1) vl = fx(vl,val[l++]);
      if(r&1) vr = fx(val[--r],vr);
      l >>= 1;
      r >>= 1;
    }
    return fx(vl,vr);
  }

  private:
    int n;
    vector<T> val;
    function<T(T,T)> fx;
    const T ex;
};